src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+
+        :param data: Diagonal matrix entries.
+        :param dtype: Datatype. Default of `None` will attempt
+            passthrough of `data.dtype`.  When explicitly provided, will
+            attempt casting `data` as needed.
+        :return: `DiagMatrix` instance.
+        """
+
+        # Assign the datatype.
+        if dtype is None:
+            dtype = data.dtype
+        self.dtype = np.dtype(dtype)
+
+        # Assign the `data`
+        self._data = data.astype(self.dtype, copy=False)
+
+        # Assign shapes from `data`
+        self.count = self._data.shape[-1]
+        self.stack_shape = self._data.shape[:-1]
+        self.shape = self._data.shape
+        # Total number of stack elements
+        self.size = np.prod(self.stack_shape)
+
+        # Numpy interop
+        # https://numpy.org/devdocs/user/basics.interoperability.html#the-array-interface-protocol
+        self.__array_interface__ = self.asnumpy().__array_interface__
+        self.__array__ = self.asnumpy()
+
+    def stack_reshape(self, *args):
+        """
+        Reshape the stack axis.
+
+        :*args: Integer(s) or tuple describing the intended shape.
+
+        :returns: `DiagMatrix` instance.
+        """
+
+        # If we're passed a tuple, use that
+        if len(args) == 1 and isinstance(args[0], tuple):
+            shape = args[0]
+        else:
+            # Otherwise use the variadic args
+            shape = args
+
+        # Sanity check the size
+        if shape != (-1,) and np.prod(shape) != self.size:
+            raise ValueError(
+                f"Number of images {self.size} cannot be reshaped to {shape}."
+            )
+
+        return DiagMatrix(self._data.reshape(*shape, self._data.shape[-1]))
+
+    def asnumpy(self):
+        """
+        Return data as Numpy array.
+
+        Note this is a read-only view.
+
+        :return: Numpy array of `self.dtype`.
+        """
+
+        view = self._data.view()
+        view.flags.writeable = False
+        return view
+
+    def __repr__(self):
+        """
+        String represention describing instance.
+        """
+        return "DiagMatrix({}, {})".format(repr(self._data), repr(self.dtype))
+
+    def copy(self):
+        """
+        Returns new `DiagMatrix` which is a copy of `self`.
+
+        :return `DiagMatrix` like self
+        """
+
+        return DiagMatrix(self._data.copy())
+
+    def __getitem__(self, key):
+        """
+        Convenience wrapper, getter on self._data.
+        """
+
+        return self._data[key]
+
+    def __len__(self):
+        """
+        Convenience function for getting `n`.
+        """
+
+        return self.count
+
+    def __check_compatible(self, other):
+        """
+        Sanity check two `DiagMatrix` instances are compatible
+        """
+
+        self.__check_size_compatible(other)
+        self.__check_dtype_compatible(other)
+
+    def __check_size_compatible(self, other):
```

Comment:
This doesn't check for broadcast compatibility of the stack shape. Do we want this or do we expect this to be handled elsewhere (such as in the `self._data + other` call below in `add`)?

Reply:
Oof, this was a little sloppy on my part. Sorry! I added it halfway through and forget to implement it consistently. I see now that I need to add some interop cases for `add` and `sub`.  We didn't use them in my basis branch, but I should have them for completeness (and I might need to use them now following some of the other changes done in this branch).

Sorry I didn't catch it in my self review. I will take care of it and respond with the updates.

With that said, if I'm going to check the stack axis it should probably only be for Diag-Diag operations at this time. (Blk does not have a stack and I don't think we want to encourage the ndarray cases when avoidable).

Reply:
Added the broadcast stack shape check to `__check_size`.

Added the missing interop cases and four tests to cover those.

Reply:
Cool!